Userland/Services/WindowServer/Compositor.cpp

```diff
@@ -1144,13 +1146,19 @@ void Compositor::recompute_overlay_rects()
     // regular window contents. This effectively just forces those areas to
     // be rendered as transparency areas, which allows us to render these
     // flicker-free.
+    swap(m_last_rendered_overlay_rects, m_overlay_rects);
     m_overlay_rects.clear_with_capacity();
     for (auto& overlay : m_overlay_list) {
         auto& render_rect = overlay.rect();
         m_overlay_rects.add(render_rect);
 
+        // Invalidate areas that are no longer in the rendered area because the overlay was moved.
+        auto previous_rects = overlay.current_render_rect().shatter(render_rect);
+        for (auto& rect : previous_rects)
+            invalidate_screen(rect);
+
         // Save the rectangle we are using for rendering from now on
-        overlay.did_recompute_occlusions();
+        bool was_invalidated = overlay.apply_render_rect();
```

Comment:
nit: `was_invalidated` can mean "this just got invalidated" or "this was in the invalidated state before this call". I think most readers will understand it to have the first meaning, while (if I understand things right) it actually has the second meaning. Maybe `needs_invalidation` is a better name? (If so, also in the impl of `apply_render_rect`)